library/ssl_tls.c

```diff
@@ -9926,4 +9927,117 @@ int mbedtls_ssl_verify_certificate(mbedtls_ssl_context *ssl,
+        ret = MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+        goto exit;
+    }
+
+    prf_input = mbedtls_calloc(prf_input_len, sizeof(unsigned char));
+    label_str = mbedtls_calloc(label_len + 1, sizeof(char));
+    if (prf_input == NULL || label_str == NULL) {
+        ret = MBEDTLS_ERR_SSL_ALLOC_FAILED;
+        goto exit;
+    }
+
+    memcpy(label_str, label, label_len);
+    label_str[label_len] = '\0';
+
+    /* The input to the PRF is client_random, then server_random.
+     * If a context is provided, this is then followed by the context length
+     * as a 16-bit big-endian integer, and then the context itself. */
+    memcpy(prf_input, ssl->transform->randbytes + 32, 32);
+    memcpy(prf_input + 32, ssl->transform->randbytes, 32);
+    if (use_context) {
+        prf_input[64] = (unsigned char) ((context_len >> 8) & 0xff);
+        prf_input[65] = (unsigned char) (context_len & 0xff);
+        memcpy(prf_input + 66, context, context_len);
+    }
+    ret = tls_prf_generic(hash_alg, ssl->session->master, 48, label_str,
+                          prf_input, prf_input_len,
+                          out, key_len);
+
+exit:
+    mbedtls_free(prf_input);
+    mbedtls_free(label_str);
+    return ret;
+}
+#endif
+
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3)
+static int mbedtls_ssl_tls13_export_keying_material(mbedtls_ssl_context *ssl,
+                                                    const mbedtls_md_type_t hash_alg,
+                                                    uint8_t *out,
+                                                    const size_t key_len,
+                                                    const char *label,
+                                                    const size_t label_len,
+                                                    const unsigned char *context,
+                                                    const size_t context_len)
+{
+    const psa_algorithm_t psa_hash_alg = mbedtls_md_psa_alg_from_type(hash_alg);
+    const size_t hash_len = PSA_HASH_LENGTH(hash_alg);
+    const unsigned char *secret = ssl->session->app_secrets.exporter_master_secret;
+
+    if (key_len > 0xffff || label_len > 250) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
+    }
+
+    return mbedtls_ssl_tls13_exporter(psa_hash_alg, secret, hash_len,
+                                      (const unsigned char *) label, label_len,
+                                      context, context_len, out, key_len);
+}
+#endif
+
+int mbedtls_ssl_export_keying_material(mbedtls_ssl_context *ssl,
+                                       uint8_t *out, const size_t key_len,
+                                       const char *label, const size_t label_len,
+                                       const unsigned char *context, const size_t context_len,
+                                       const int use_context)
+{
+    if (!mbedtls_ssl_is_handshake_over(ssl)) {
+        return MBEDTLS_ERR_SSL_BAD_INPUT_DATA;
```

Comment:
I've had a look around and it seems like `MBEDTLS_ERR_SSL_BAD_INPUT_DATA` being used like this has precedent [here](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L5353) and [here](https://github.com/Mbed-TLS/mbedtls/blob/development/include/mbedtls/ssl.h#L5435). In both of these cases the error is used to signal that the caller should wait for the handshake to finish.

With that in mind I'm happy to follow existing convention and leave this as-is. Since it's a bit unusual, could you specifically document this case in the docstring? Something like:
```
\return         0 on success, or
                MBEDTLS_ERR_BAD_INPUT_DATA, if the handshake has not completed yet, or
                Another SSL specific error in case of a different failure.
```

Reply:
@davidhorstmann-arm I share your feeling that `BAD_INPUT_DATA` is not the right error code here. It would be ok for an internal function where it's a coding error to ever invoke this function without knowing the handshake state, but it feels icky in TLS export, especially since it could be called unknowingly during renegotiation.

I don't see another error code that looks really appropriate. Maybe we should create one, and move the other cases you found to that error code in 4.0?

Reply:
> I don't see another error code that looks really appropriate. Maybe we should create one, and move the other cases you found to that error code in 4.0?

@gilles-peskine-arm Do you mean I should add the error code in this merge request, and you move the other cases to it later? Or should I still use `BAD_INPUT_DATA`, and you change it later?